Fix file not closing

DuzaBF opened the pull request:
This fixes Issue #17

MicheleCotrufo commented:
Thank you for spotting this bug!
I initially wanted to structure the code similarly to what you suggested (with the `with open(file, 'rb') as file:`  statement). The only issue is that the input argument of the function `pdf2doi_singlefile` could be either a string (containing the path of the file) or a file object already opened by the user. This second scenario is useful when the `pdf2doi` library is used within another python script, and the user wants to find the DOI of a pdf file already opened (see discussion here https://github.com/MicheleCotrufo/pdf2doi/issues/11). Thus, when the input argument is not a string, the function `pdf2doi_singlefile` should treat it as a file object (assuming that it is already opened) and not close the file, since the user will take care of it in his/her own script.
In the current version, I check if the input argument is a string, and if not I open the file. However, as you currently pointed out, the file doesn't get closed and this is a problem for `pdf-renamer`. 
I could probably add the `if flag_closefile: file.close()` every time before calling `return result`, but it would look quite ugly. Do you have any suggestions on how to reshape the code?

MicheleCotrufo commented:
I guess a possible semi-elegant solution would be to go back to previous version of code (i.e. before your pull request), create an auxiliary function (e.g. `_pdf2doi_singlefile`) that contains the code

    #First method: we look into the pdf metadata (in the current implementation this is done
    # via the getDocumentInfo() method of the library PyPdf) and see if any of them is a string which containts a
    #valid identifier inside it. We first look for the elements of the dictionary with keys '/doi' or /identifier'(if the they exist), 
    #and then any other field of the dictionary 
    logger.info(f"Method #1: Looking for a valid identifier in the document infos...")
    result = finders.find_identifier(file,method="document_infos",keysToCheckFirst=['/doi','/identifier'])
    if result['identifier']:
        return result 

    #Second method: We look for a DOI or arxiv ID inside the filename
    logger.info(f"Method #2: Looking for a valid identifier in the file name...")
    result = finders.find_identifier(file,method="filename")
    if result['identifier']:
        return result 

    #Third method: We look in the plain text of the pdf and try to find something that matches a valid identifier. 
    logger.info(f"Method #3: Looking for a valid identifier in the document text...")
    result =  finders.find_identifier(file,method="document_text")
    if result['identifier']:
        return result 

    #Fourth method: We look for possible titles of the paper, do a google search with them, 
    # open the first results and look for identifiers in the plain text of the searcg results.
    logger.info(f"Method #4: Looking for possible publication titles...")
    result =  finders.find_identifier(file,method="title_google")
    if result['identifier']:
        return result

    #Fifth method: We extract the first N characters from the file (where N is set by config.get('N_characters_in_pdf')) and we use it as 
    # a query for a google seaerch. We open the first results and look for identifiers in the plain text of the searcg results.
    logger.info(f"Method #5: Trying to do a google search with the first {config.get('N_characters_in_pdf')} characters of this pdf file...")
    result =  finders.find_identifier(file,method="first_N_characters_google")
    if result['identifier']:
        return result

and then after the auxiliary function has returned, do the 

    if flag_closefile:
        file.close()

Thoughts?

DuzaBF commented:
I think creating a helper function would be better solution. I will add a commit in a moment.

DuzaBF commented:
Here it is. My IDE reformatted the white spaces so it looks messy, just check the last commit. I moved searching for DOI identifier to helper function and made original `pdf2doi_singlefile ` a wrapper with tries to open file and process it, and if it is unsuccessful - passes the file to helper function as it is. Also added some type hints.

MicheleCotrufo commented:
Thanks!! I did some quick tests and all seems to be working fine. I merged your commits, and I will release soon version 1.3 of pdf2doi.
